xmtp_mls/src/groups/device_sync.rs

```diff
@@ -101,57 +107,145 @@ pub enum DeviceSyncError {
+where
+    ApiClient: XmtpApi + 'static,
+    V: SmartContractSignatureVerifier + 'static,
+{
+    pub async fn start_sync_worker(
+        &self,
+        provider: &XmtpOpenMlsProvider,
+    ) -> Result<(), DeviceSyncError> {
+        self.sync_init(provider).await?;
+
+        crate::spawn(None, {
+            let client = self.clone();
+
+            let receiver = client.local_events.subscribe();
+            let mut sync_stream = Box::pin(receiver.stream_sync_messages());
+
+            async move {
+                while let Err(err) = client.sync_worker(&mut sync_stream).await {
+                    tracing::error!("Sync worker error: {err}");
+                }
+            }
+        });
+
+        Ok(())
+    }
 }
 
 impl<ApiClient, V> Client<ApiClient, V>
 where
     ApiClient: XmtpApi,
     V: SmartContractSignatureVerifier,
 {
-    pub async fn reply_to_sync_request(
+    pub(crate) async fn sync_worker(
         &self,
-        provider: &XmtpOpenMlsProvider,
-        kind: DeviceSyncKind,
-    ) -> Result<DeviceSyncReplyProto, DeviceSyncError> {
-        let conn = provider.conn_ref();
+        sync_stream: &mut Pin<Box<impl Stream<Item = Result<SyncMessage, SubscribeError>>>>,
+    ) -> Result<(), DeviceSyncError> {
+        let mut sync_stream = sync_stream.as_mut();
 
-        let (_msg, request) = self.pending_sync_request(provider, kind).await?;
-        let records = match kind {
-            DeviceSyncKind::Consent => vec![self.syncable_consent_records(conn)?],
-            DeviceSyncKind::MessageHistory => {
-                vec![self.syncable_groups(conn)?, self.syncable_messages(conn)?]
+        let provider = self.mls_provider()?;
+
+        let get_message_debounce = |conn: &DbConnection, message_id: &[u8]| {
+            let mut tries = 0;
+            while tries < 3 {
+                if let Some(msg) = conn.get_group_message(message_id)? {
+                    return Ok(msg);
+                }
+                thread::sleep(Duration::from_millis(20));
+                tries += 1;
             }
-            DeviceSyncKind::Unspecified => return Err(DeviceSyncError::UnspecifiedDeviceSyncKind),
+
+            Err(DeviceSyncError::NoPendingRequest)
         };
 
-        let reply = self
-            .create_sync_reply(&request.request_id, &records, kind)
-            .await?;
-        self.send_sync_reply(provider, reply.clone()).await?;
+        while let Some(msg) = sync_stream.next().await {
+            let msg = msg?;
+            match msg {
+                SyncMessage::Reply { message_id } => {
+                    let conn = provider.conn_ref();
+                    let msg = get_message_debounce(conn, &message_id)?;
+
+                    let msg_content: DeviceSyncContent =
+                        serde_json::from_slice(&msg.decrypted_message_bytes)?;
+                    let DeviceSyncContent::Reply(reply) = msg_content else {
+                        // This should never happen
```

Comment:
question: what if it happened? what we can do?

Reply:
Then the code is wrong. I was tempted to put an `unreachable!();` here. :wink: Thinking about it a bit more, it would probably make more sense to do that. Let me update.